### PR TITLE
Hook StaticCudaLauncher up to torch.compile (cold start)

### LINKED ISSUE #149448: Hook up StaticCudaLauncher to torch.compile

### PR DESCRIPTION
**Pull request by jamesjwu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #149629
* #149442
* #149054
* __->__ #148890

This hooks up the previous PR to torch.compile. Will add a config flag to hide this behind in a bit, but for now it's useful for testing purposes to have it on by default. 

Inductor will automatically choose to use StaticCudaLauncher to launch triton kernels if: 
- The kernel is a cuda kernel and inductor can find a cubin file associated with it
- The kernel takes less than 50 arguments
- The kernel doesn't use any special features (launch hooks, large amounts of shared memory)
- The kernel is not user defined (to be supported in a later PR)

We split CompileResult into TritonCompileResult and StaticTritonCompileResult, but have them share implementations of how they exec a python launcher. StaticTritonCompileResult's python launcher has the benefit of a simpler def_args/call_args setup, since it always filters out all constexprs before running, no matter the triton version. 

Some key features of StaticTritonCompileResult: 
- It is fully serializable
- It stores the minimum amount of stuff, so that later it can be cached easily 
- It does not depend on any triton specific types (though it does have various triton metadata). 

For now, both TritonCompileResult and StaticTritonCompileResult still `exec` custom python launchers, and use GridExpr. We can change that in the future to simplify if we'd like. For now though, this custom python codegen is good for flexibility when it comes to supporting removal of constexprs, so using it for static launching is nice to not have to pay the cost of removing constexprs at kernel runtime. 

Hooking everything up to torch.compile lets me run every unit test with StaticCudaLauncher to make sure that we still pass (even if we bypass StaticCudaLauncher itself). It also lets me check for compilation/runtime performance with these changes. 

Fixes #149448 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

**Comment by jamesjwu** [comment]
@jamesjwu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D70992220).

**Comment by jamesjwu** [comment]
@jamesjwu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D70992220).

File: torch/csrc/inductor/static_cuda_launcher.cpp [hunk 1]
```diff
@@ -311,6 +311,10 @@ PyObject* launch_kernel(PyObject* self, PyObject* args) {
           &stream)) {
     return nullptr;
   }
+  if (gridX * gridY * gridZ <= 0) {
```
Comment (jamesjwu) [inline torch/csrc/inductor/static_cuda_launcher.cpp:314]: Found this bug from running unit tests with StaticCudaLauncher: triton ignores all calls to kernels when outside of grid boundaries

File: test/inductor/test_static_cuda_launcher.py [hunk 2]
```diff
@@ -1,13 +1,12 @@
 # Owner(s): ["module: inductor"]
```
Comment (jamesjwu) [inline test/inductor/test_static_cuda_launcher.py:1]: I went ahead and revamped the testing setup so we don't have to manually call save_to_tmp_file() each time, instead opting to set _cubin_path similar to how `triton_heuristics` handles it. I think it's cleaner this way. No changes in behavior to existing tests.

**Comment by jamesjwu** [comment]
Kicked off a new set of benchmarks, will paste here when done.

**Comment by jamesjwu** [comment]
https://hud.pytorch.org/benchmark/torchbench/inductor_with_cudagraphs?dashboard=torchinductor&startTime=Mon,%2010%20Mar%202025%2023:58:14%20GMT&stopTime=Mon,%2017%20Mar%202025%2023:58:14%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(a100)&lBranch=gh/jamesjwu/119/head&lCommit=219edea41e40aeff088584f11cb1ecc03b54b81d&rBranch=gh/jamesjwu/119/base&rCommit=82a89371229e09fc5400fd9f5e90161492d501d7

Initial benchmarks look good, will run with H100 as well

File: torch/_inductor/runtime/static_cuda_launcher.py [hunk 3]
```diff
@@ -2,13 +2,11 @@
...
 from typing_extensions import Unpack
 
-from torch.utils._ordered_set import OrderedSet
-
 from .triton_compat import ASTSource, CompiledKernel
 
 
 MAX_SHARED_MEMORY = 49152
-MAX_ARGS = 50
+MAX_ARGS = 120
```
Comment (jamesjwu) [inline torch/_inductor/runtime/static_cuda_launcher.py:9]: I'm confirming that increasing MAX_ARGS does not cause OOMs or affect kernel performance. 120 args * 8 bytes per args = 0.96 kb, which should be basically nothing.

**Review by jansel — CHANGES_REQUESTED** [review]
1) Looks like there are some new failure (and nice speedups!) in that bechmark run, we should fix the failures.

2) Can we add support for larger number of args (fall back to dynamic alloc), and the special stuff like hooks, shared memory, custom ops?  I would really like to have only one code path since I think that will be easier to maintain.   I suspect stuff like args>50 will have poor coverage in our test suite and is likey to get broken without us noticing.

**Comment by jamesjwu** [comment]
> Can we add support for larger number of args (fall back to dynamic alloc), and the special stuff like hooks, shared memory, custom ops? I would really like to have only one code path since I think that will be easier to maintain. I suspect stuff like args>50 will have poor coverage in our test suite and is likey to get broken without us noticing.

I'm happy to implement these additions so that our coverage is as complete as possible, but I am very worried about trying to ship this without a config flag to turn it off initially or fall back on (especially internally, where rollouts can take weeks). So I think we should still be able to support the fallback to triton for now. I do agree directionally that we should try to get to the case where all kernels are statically launchable. 

Re: falling back to dynamic alloc: what do you think about falling back to something like `alloca`, to just allocate it directly on the stack? I know it might not be super portable, but maybe it's possible to support. 

I fixed a bug just now (when we have a global scratch, MAX_ARGS on C++ side actually needs to be MAX_ARGS + 1 lol), will kick off a new benchmark run.

**Comment by jamesjwu** [comment]
I realize there's a big list of features that we technically need to support to be "feature parity" with triton's launcher capabilities, so I wrote down them all in this issue: https://github.com/pytorch/pytorch/issues/149440, and will fix them one by one: https://github.com/pytorch/pytorch/pull/149442.

The goal is that with `config.static_cuda_launch=True`, all triton kernels are statically launched (or else we hard error). This should decrease the complexity of maintaining the code base, because then either all of the CompileResults are triton launched, or all of them are statically launched, but there's no in between. 

@jansel I do have a concern, though, with user defined triton kernels, because they can essentially contain *anything*. So there's a bunch of features that PT2 generated kernels will never encounter (i.e. `num_ctas>1`, `CUTensorMaps`, etc), but would need to be implemented in the static launcher just to support user defined ones. We could fix the finite list of these today, but the bigger worry is that until we can fully upstream it to triton side, there's always a risk of triton supporting a new feature that StaticCudaLauncher can't support, and then having user defined triton kernels break. What do you think our options are here? Will we need to keep a fallback option around for those?

**Comment by jamesjwu** [comment]
- Fixed a bunch of bugs and rebased onto main
- Ran a new benchmark:
https://hud.pytorch.org/benchmark/huggingface/inductor_with_cudagraphs?dashboard=torchinductor&startTime=Wed%2C%2012%20Mar%202025%2015%3A18%3A30%20GMT&stopTime=Wed%2C%2019%20Mar%202025%2015%3A18%3A30%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(a100)&lBranch=gh/jamesjwu/119/head&lCommit=be6d18d1307e269d4dac6594f25dae1941787e69&rBranch=gh/jamesjwu/119/base&rCommit=f461654bc928f8af1580b6ce1a13635af99e3ab3

(For posterity, base revision is f461654bc928f8af1580b6ce1a13635af99e3ab3 and head revision is be6d18d1307e269d4dac6594f25dae1941787e69)

Overall looks better, no more accuracy failures or fail to runs. One model, XGLMForCausalLM, seemed to have regressed runtime perf, but I couldn't repro it locally. I checked on various other revisions (i.e. main, other base revisions I've run), and I think this particular base revision might have just gotten a high perf speedup due to variance (2.5x), though I will rerun the head revision one more time to confirm.

**Comment by jamesjwu** [comment]
Ugh there's one more fail_accuracy on timm_efficientnet on torchbench. But I checked on main revisions, and it seems to have been failing accuracy until today? 
image

**Comment by jansel** [comment]
> Re: falling back to dynamic alloc: what do you think about falling back to something like alloca, to just allocate it directly on the stack? I know it might not be super portable, but maybe it's possible to support.

This seems fine to me.

For user-defined Triton kernels maybe we just always fallback to the upstream launcher?  That way the codepath will at least be well tested.

**Comment by jamesjwu** [comment]
> For user-defined Triton kernels maybe we just always fallback to the upstream launcher? That way the codepath will at least be well tested.

This seems good to me, I had thought your point was to try to get it so we don't have to keep both codepaths. But we can make it so only user defined triton kernels use the old launcher system and everything else uses static launch.

**Comment by jamesjwu** [comment]
One last round of benchmarks, showing stuff looks good (this is before I flipped the config to default off):

https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Thu%2C%2013%20Mar%202025%2015%3A39%3A38%20GMT&stopTime=Thu%2C%2020%20Mar%202025%2015%3A39%3A38%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cuda%20(a100)&lBranch=gh/jamesjwu/119/orig&lCommit=4673097e21a35158bb28402e35af65f620846f9f&rBranch=main&rCommit=e8a35eb7da1f7cbd62827943c1e00849a8592298

Will flip the config to default on once I land https://github.com/pytorch/pytorch/pull/149442 and some other in progress PRs to reach closer feature parity with triton's launcher.

**Review by jansel — APPROVED** [review]
